### ISSUE TITLE
How to solve slow computation speed of CustomNonbondedForce interaction

### ISSUE THREAD

**Issue by arjunwadhawan** [pandegroup/openmm#2124]
Hey,

I am using CustomNonbondedForce to compute the interaction in my system. The speed of computation is ~30-40 % slower when using CustomNonbondedForce compared to standard way. I hope this is an expected behaviour. 

Is there any possible way to increase the simulation speed when using CustomNonbondedForce ? (or is there something that I am doing wrong)
.
My goal is to use CustomNonbondedForce as it will allow me to scale the interactions or change the combination rule (etc..) 

Attached is a test system to check both CustomNonbondedForce and  standard way.. 
[CustomNonbondedForce_speed.zip](https://github.com/pandegroup/openmm/files/2209807/CustomNonbondedForce_speed.zip)
To run the system , use the following inside the directory

`python *.py `

_Simulation Details-_
System - NaCl + Water 
Openmm version-  7.1.1
Openmm tools version - 0.14.0
Python version- Python 3.5.4
The interaction is calculated by Lorentz- Berthelot combination rule.  

Let me know if any other information is required

**Comment by peastman** [comment]
You're comparing two very different systems.  You start by calling `forcefield.createSystem()` to create a System that includes a NonbondedForce.  You then add six different CustomNonbondedForces to it.  So you're looking at a system with a total of seven nonbonded forces (custom and standard) in it.  Of course that's slower than a system that only has one.

So there are two issues here.  First, you need to remove the NonbondedForce.  The custom forces are supposed to replace it, not be in addition to it.  Second, using lots of identical CustomNonbondedForces is very inefficient.  You should use only one, relying on the per-particle parameters to distinguish between particle types.

**Comment by arjunwadhawan** [comment]
Thanks for the reply @peastman . I understand what you said. 

1- remove the NonbondedForce: I can remove it by using `system.removeForce` 
2- identical CustomNonbondedForces : I do not know how to execute this. Can you give an example of how to write this ? I tried it with following but get an error 

```
def nacl_madrid_ff_3(system, topology):
    cutoff =1 * nanometer
    
    # Determine form interaction groups. These are the index of oxygen from H2O and Co2
    na_atoms = set()
    cl_atoms = set()
    oxygen_atoms = set()
    all_relevent_atoms = set()
    
    
    for atom in topology.atoms():
        if atom.name in ['Na']:
            na_atoms.add(atom.index)
        if atom.name in ['Cl']:
            cl_atoms.add(atom.index)
        if atom.name in ['O']:
            oxygen_atoms.add(atom.index)
        if atom.name in ['Na', 'Cl', 'O']:
            all_relevent_atoms.add(atom.index)
            
        
    # Get all the forces in the sytem 
    forces = {system.getForce(index).__class__.__name__: system.getForce(index) for index in range(system.getNumForces())}
    
    # Select the Non bonded forces from 'forces' dict
    nonbonded_force = forces['NonbondedForce']
    
    
# Overall interaction 
    energy_expression = '4*epsilon*((sigma/r)^12-(sigma/r)^6); sigma=sqrt(sigma1*sigma2); epsilon=sqrt(epsilon1*epsilon2)'
   
    custom_force  =  CustomNonbondedForce(energy_expression)
    custom_force.setNonbondedMethod(CustomNonbondedForce.CutoffPeriodic)
    
    custom_force.addPerParticleParameter('sigma')
    custom_force.addPerParticleParameter('epsilon')
    custom_force.setCutoffDistance(nonbonded_force.getCutoffDistance())
    

    for index in range(system.getNumParticles()):

        if index in na_atoms:
            custom_force.addParticle([0.221737, 1.47236])
            custom_force.addInteractionGroup(na_atoms, na_atoms) # Na - Na 
        if index in cl_atoms:
            custom_force.addParticle([0.484906, 0.076923])
            custom_force.addInteractionGroup(cl_atoms, cl_atoms) # Cl - Cl
        if index in cl_atoms or na_atoms:
            custom_force.addParticle([0.290512, 1.43889])
            custom_force.addInteractionGroup(na_atoms, cl_atoms) # Na - Cl
        if index in oxygen_atoms:
            custom_force.addParticle([0.315890, 0.774907])   
            custom_force.addInteractionGroup(oxygen_atoms, oxygen_atoms) # O - O
        if index in na_atoms or oxygen_atoms:
            custom_force.addParticle([0.251338, 0.793388])
            custom_force.addInteractionGroup(na_atoms, oxygen_atoms) # Na - O
        if index in cl_atoms or oxygen_atoms:
            custom_force.addParticle([0.426867, 0.061983])
            custom_force.addInteractionGroup(cl_atoms, oxygen_atoms)  # Cl - O
    system.addForce(custom_force) 
    
    return system
```

Error : `Exception: CustomNonbondedForce must have exactly as many particles as the System it belongs to.`

**Comment by Lnaden** [comment]
It looks like you are adding the interactions between all heavy atoms, but are never adding your Hydrogen atoms to the system, so you only add in about 3000 heavy atoms and miss the 3000 water atoms, so that causes the particle mismatch.

Secondly, the `addInteractionGroup` does not behave the way you appear to be using it. When you add a particle to a `CustomNonbondedForce`, it is assumed that every particle interacts with every other particle, just like in a standard `NonbondedForce`. If you specify an Interaction Group however, it tells the force to *only* compute interactions between the particles in argument 1 with the particles in argument 2. so `addInteractionGroup(all_particles, all_particles)` would mechanically be identical to not having any interaction groups at all (although redundant). However, if you specify `addInteractionGroup(na_atoms, oxygen_atoms)` then *only* the sodium-oxygen interaction would be considered. Adding more interaction groups increases the pairs of atoms which are included, but does not make the union. e.g. 

```python 
custom_force.addInteractionGroup(na_atoms, oxygen_atoms)
custom_force.addInteractionGroup(oxygen_atoms, cl_atoms)
```
creates 2 sets of interactions between Na-O and O-Cl, but *not* Na-Cl*.

With all that said, should you choose to use interaction groups, they only have to be invoked once, not once per added atom. This goes back to the way the default CustomNonbondedForce behaves, when you add multiple atoms, its automatically inferred that it will interact with all atoms added before and after. 

The previous point leads to something else I see, it looks like you are trying to re-add atoms multiple times, for instance with your `cl_atoms or na_atoms` you add a particle you have already added from the `cl_atoms` alone or the `na_atoms` alone, so you are adding by the end something like 20000 particles to the system (and 20000 interaction groups).

So to summarize:
* All particles added to the CustomNonbondedForce are assumed to interact with every other particle by default
* You have to add all atoms, not just heavy ones
* You only add each particle once, specifying its parameters (sigma/epsilon here) and let the mixing rules defined in the energy expression handle how pairs interact
* Interaction groups define subsets of interactions to compute, if desired
* Interaction groups only need to be defined once per *pair of sets*, not pair of particles

Hope this helps!

**Comment by arjunwadhawan** [comment]
Thanks for helping @Lnaden . I realized my mistakes. 

1- Firstly, I do not need mixing rule. The interaction values of each pair (sigma and epsilon) is given in the table. (Like σ, ε for Na Na interaction is sigma = 0.221737 ; epsilon = 1.47236). These values are based on charge scaling in the original paper. The σ, ε for individual atom is not given. 

2- I would need to include both steric and electrostatic forces in CustomNonbondedForce (when I am removing the `forces['NonbondedForce']` from my system)

I tried to implement it as below. I have 2 questions :-

Question 1- The speed of the simulation when using CustomNonbondedForce is still 50 % slower than the standard way of calculating NonbondedForce (same system). In the standard way, I use the  σ, ε values of each particle based on another article where σ, ε for the individual atoms is give but these are old and dont match with experiments (A. L. Benavides, J. L. Aragones, and C. Vega, J. Chem. Phys. 144, 124504) .  Attached is a test system. Is there anything wrong with my implementation or would there be a way to increase the calculation speed?
[CustomNonbondedForce_speed_2.zip](https://github.com/pandegroup/openmm/files/2228315/CustomNonbondedForce_speed_2.zip)

`def nacl_madrid_ff_single_expression(system, topology):
    cutoff =1 * nanometer
    
    # Determine form interaction groups. These are the index of oxygen from H2O and NaCl
    na_atoms = set()
    cl_atoms = set()
    oxygen_atoms = set()
    
    for atom in topology.atoms():
        if atom.name in ['Na']:
            na_atoms.add(atom.index)
        if atom.name in ['Cl']:
            cl_atoms.add(atom.index)
        if atom.name in ['O']:
            oxygen_atoms.add(atom.index)

    # Get all the forces in the sytem 
    forces = {system.getForce(index).__class__.__name__: system.getForce(index) for index in range(system.getNumForces())}
    
    # Select the Non bonded forces from 'forces' dict
    nonbonded_force = forces['NonbondedForce']
    
    
###-------------------------------------------###
# Steric (van der waals) interactions 
###-------------------------------------------###
    
#  Nonbonded force expression 
    expression_total = "4*epsilon*((sigma/r)^12-(sigma/r)^6);"

# Calculate Sigma 
    expression_total += "sigma = (sigma_nacl1+sigma_nacl2)/2 + (sigma_nao1+sigma_nao2)/2 + (sigma_clo1+sigma_clo2)/2 -"
    expression_total += "(sigma_nana1+sigma_nana2)/2 - (sigma_clcl1+sigma_clcl2)/2 - (sigma_oo1+sigma_oo2)/2;"
    
# Calculate epsilon
    expression_total += "epsilon =  (epsilon_nacl1+epsilon_nacl2)/2+ (epsilon_nao1+epsilon_nao2)/2 + (epsilon_clo1+epsilon_clo2)/2 -"
    expression_total += "(epsilon_nana1+epsilon_nana2)/2 - (epsilon_clcl1+epsilon_clcl2)/2 -(epsilon_oo1+epsilon_oo2)/2;"
    
    
    
    total_interaction_force =  CustomNonbondedForce(expression_total)
    total_interaction_force.setNonbondedMethod(CustomNonbondedForce.CutoffPeriodic)
    total_interaction_force.addPerParticleParameter('sigma_nana')
    total_interaction_force.addPerParticleParameter('sigma_clcl')
    total_interaction_force.addPerParticleParameter('sigma_nacl')
    total_interaction_force.addPerParticleParameter('sigma_oo')
    total_interaction_force.addPerParticleParameter('sigma_nao')
    total_interaction_force.addPerParticleParameter('sigma_clo')

    total_interaction_force.addPerParticleParameter('epsilon_nana')
    total_interaction_force.addPerParticleParameter('epsilon_clcl')
    total_interaction_force.addPerParticleParameter('epsilon_nacl')
    total_interaction_force.addPerParticleParameter('epsilon_oo')
    total_interaction_force.addPerParticleParameter('epsilon_nao')
    total_interaction_force.addPerParticleParameter('epsilon_clo')

    total_interaction_force.setCutoffDistance(nonbonded_force.getCutoffDistance())
   
    for index in range(system.getNumParticles()):
        # Define Sigma(s) and epsilon(s) for Na : These are scaled values. Final value is calculated in expression 
        if index in na_atoms:
            total_interaction_force.addParticle([0.320, 0, 0.290, 0, 0.251, 0,    0.75995, 0, 1.43887, 0, 0.793388, 0])

        # Define Sigma(s) and epsilon(s) for Cl :These are scaled values. Final value is calculated in expression 
        elif index in cl_atoms:
            total_interaction_force.addParticle( [0, 0.232473, 0.290512, 0, 0, 0.42686,   0, 1.42395, 1.43887, 0, 0, 0.061983])
            
        # Define Sigma(s) and epsilon(s) for Oxygen : These are scaled values. Final value is calculated in expression 
        elif index in oxygen_atoms:
             total_interaction_force.addParticle( [0,0,0, 0.362, 0.251, 0.426,   0,0,0, 0.080464, 0.793388, 0.061983 ])

        # Rest all is 0 (as Hydrogens and M site of water do not interact)
        else:
            total_interaction_force.addParticle([0, 0, 0 ,0, 0, 0, 0 ,0 , 0, 0, 0 ,0 ])
    
    for index in range(nonbonded_force.getNumExceptions()):
        [iatom, jatom, chargeprod, sigma, epsilon]  = nonbonded_force.getExceptionParameters(index)
        total_interaction_force.addExclusion(iatom,jatom )    
    

    system.addForce(total_interaction_force)
    
###-------------------------------------------###
# Electrostatics (columbic) interactions 
###-------------------------------------------###
    
    
     # This bit is for calculating the smooth PME based electrostatic 
    [alpha_ewald, nx, ny, nz] = nonbonded_force.getPMEParameters()
    if (alpha_ewald/alpha_ewald.unit) == 0.0:
    # If alpha is 0.0, alpha_ewald is computed by OpenMM from from the error tolerance.
        tol = nonbonded_force.getEwaldErrorTolerance()
        alpha_ewald = (1.0/nonbonded_force.getCutoffDistance()) * np.sqrt(-np.log(2.0*tol))

    expression_columbic  = "(ONE_4PI_EPS0*chargeprod*erfc(alpha_ewald*r)/r);"
    expression_columbic += "ONE_4PI_EPS0 = {:f};".format(ONE_4PI_EPS0)  # already in OpenMM units
    expression_columbic += "chargeprod = charge1*charge2;"
    expression_columbic += 'alpha_ewald = %.16e;' % (alpha_ewald.value_in_unit_system(unit.md_unit_system))
    columbic_force = CustomNonbondedForce(expression_columbic)
    columbic_force.addPerParticleParameter("charge")
    columbic_force.setNonbondedMethod(CustomNonbondedForce.CutoffPeriodic)
    
    for index in range(nonbonded_force.getNumParticles()):
        [charge, sigma, epsilon] = nonbonded_force.getParticleParameters(index)
        columbic_force.addParticle([charge])
        # Dont add interaction group, it will atomatically add for all the particles 

    for index in range(nonbonded_force.getNumExceptions()):
        [iatom, jatom, chargeprod, sigma, epsilon]  = nonbonded_force.getExceptionParameters(index)
        columbic_force.addExclusion(iatom,jatom )
    
    system.addForce(columbic_force)

    return system
 `


Question 2- As @peastman suggested to use less identical expressions for CustomNonbondedForces , I wanted to compare the speed of using 1 expression vs 6 expressions for the steric force. It looks like that the speed of the simulation is almost the same in both cases. The script is below. Again, do you find anything that I may be doing wrong here? Attached is a test system.

`# This version has steric and electrostatic seperate 

def nacl_madrid_ff_six_expression(system, topology):
    cutoff =1 * nanometer
    
    # Determine form interaction groups. These are the index of oxygen from H2O and Co2
    na_atoms = set()
    cl_atoms = set()
    oxygen_atoms = set()

    
    for atom in topology.atoms():
        if atom.name in ['Na']:
            na_atoms.add(atom.index)
        if atom.name in ['Cl']:
            cl_atoms.add(atom.index)
        if atom.name in ['O']:
            oxygen_atoms.add(atom.index)

        
    # Get all the forces in the sytem 
    forces = {system.getForce(index).__class__.__name__: system.getForce(index) for index in range(system.getNumForces())}
    
    # Select the Non bonded forces from 'forces' dict
    nonbonded_force = forces['NonbondedForce']

    # This bit is for calculating the smooth PME based electrostatic 
    [alpha_ewald, nx, ny, nz] = nonbonded_force.getPMEParameters()
    if (alpha_ewald/alpha_ewald.unit) == 0.0:
    # If alpha is 0.0, alpha_ewald is computed by OpenMM from from the error tolerance.
        tol = nonbonded_force.getEwaldErrorTolerance()
        alpha_ewald = (1.0/nonbonded_force.getCutoffDistance()) * np.sqrt(-np.log(2.0*tol))

###-------------------------------------------###
# Steric (van der waals) interactions 
###-------------------------------------------###

# Na - Na interaction 
    expression_na_na = '4*epsilon*((sigma/r)^12-(sigma/r)^6) ;'  
    expression_na_na += 'sigma = 0.221737 ; epsilon = 1.47236;'

    na_na_interaction_force =  CustomNonbondedForce(expression_na_na)
    na_na_interaction_force.setNonbondedMethod(CustomNonbondedForce.CutoffPeriodic)
    na_na_interaction_force.setUseSwitchingFunction(False)
    na_na_interaction_force.setCutoffDistance(nonbonded_force.getCutoffDistance())
    na_na_interaction_force.addInteractionGroup(na_atoms, na_atoms)

    system.addForce(na_na_interaction_force)
    
# Cl - Cl interaction 
    
    expression_cl_cl = '4*epsilon*((sigma/r)^12-(sigma/r)^6);'
    expression_cl_cl += 'sigma =0.484906 ; epsilon = 0.076923;'
    
    cl_cl_interaction_force = CustomNonbondedForce(expression_cl_cl)
    cl_cl_interaction_force.setNonbondedMethod(CustomNonbondedForce.CutoffPeriodic)
    cl_cl_interaction_force.setUseSwitchingFunction(False)
    cl_cl_interaction_force.setCutoffDistance(nonbonded_force.getCutoffDistance())
    cl_cl_interaction_force.addInteractionGroup(cl_atoms, cl_atoms)

    system.addForce(cl_cl_interaction_force)    
    
# Na - Cl interaction 

    expression_na_cl = '4*epsilon*((sigma/r)^12-(sigma/r)^6);'
    expression_na_cl += 'sigma = 0.290512 ; epsilon = 1.43889;'
    
    na_cl_interaction_force = CustomNonbondedForce(expression_na_cl)
    na_cl_interaction_force.setNonbondedMethod(CustomNonbondedForce.CutoffPeriodic)
    na_cl_interaction_force.setUseSwitchingFunction(False)
    na_cl_interaction_force.setCutoffDistance(nonbonded_force.getCutoffDistance())
    na_cl_interaction_force.addInteractionGroup(na_atoms, cl_atoms)
    
    system.addForce(na_cl_interaction_force) 

# O - O interaction 

    expression_o_o = '4*epsilon*((sigma/r)^12-(sigma/r)^6);'
    expression_o_o += 'sigma = 0.315890; epsilon= 0.774907;'
    
    o_o_interaction_force = CustomNonbondedForce(expression_o_o)
    o_o_interaction_force.setNonbondedMethod(CustomNonbondedForce.CutoffPeriodic)
    o_o_interaction_force.setUseSwitchingFunction(False)
    o_o_interaction_force.setCutoffDistance(nonbonded_force.getCutoffDistance())
    o_o_interaction_force.addInteractionGroup(oxygen_atoms, oxygen_atoms)

    system.addForce(o_o_interaction_force) 

# Na - O interaction 
    
    expression_na_o = '4*epsilon*((sigma/r)^12-(sigma/r)^6);'
    expression_na_o += 'sigma = 0.251338 ; epsilon = 0.793388;'
    
    na_o_interaction_force = CustomNonbondedForce(expression_na_o)
    na_o_interaction_force.setNonbondedMethod(CustomNonbondedForce.CutoffPeriodic)
    na_o_interaction_force.setUseSwitchingFunction(False)
    na_o_interaction_force.setCutoffDistance(nonbonded_force.getCutoffDistance())
    na_o_interaction_force.addInteractionGroup(na_atoms, oxygen_atoms)

    system.addForce(na_o_interaction_force) 
    
# Cl - O interaction 
    
    expression_cl_o = '4*epsilon*((sigma/r)^12-(sigma/r)^6);'
    expression_cl_o += 'sigma = 0.426867 ; epsilon = 0.061983;'
    
    cl_o_interaction_force = CustomNonbondedForce(expression_cl_o)
    cl_o_interaction_force.setNonbondedMethod(CustomNonbondedForce.CutoffPeriodic)
    cl_o_interaction_force.setUseSwitchingFunction(False)
    cl_o_interaction_force.setCutoffDistance(nonbonded_force.getCutoffDistance())
    cl_o_interaction_force.addInteractionGroup(cl_atoms, oxygen_atoms)
    
    system.addForce(cl_o_interaction_force) 

###-------------------------------------------###
# Electrostatics (columbic) interactions 
###-------------------------------------------###


# columbic  interaction 

    expression_columbic  = "(ONE_4PI_EPS0*chargeprod*erfc(alpha_ewald*r)/r);"
    expression_columbic += "ONE_4PI_EPS0 = {:f};".format(ONE_4PI_EPS0)  # already in OpenMM units
    expression_columbic += "chargeprod = charge1*charge2;"
    expression_columbic += 'alpha_ewald = %.16e;' % (alpha_ewald.value_in_unit_system(unit.md_unit_system))
    columbic_force = CustomNonbondedForce(expression_columbic)
    columbic_force.addPerParticleParameter("charge")
    columbic_force.setNonbondedMethod(CustomNonbondedForce.CutoffPeriodic)
    
    for index in range(nonbonded_force.getNumParticles()):
        [charge, sigma, epsilon] = nonbonded_force.getParticleParameters(index)
        columbic_force.addParticle([charge])
        # Dont add interaction group, it will atomatically add for all the particles 

    for index in range(nonbonded_force.getNumExceptions()):
        [iatom, jatom, chargeprod, sigma, epsilon]  = nonbonded_force.getExceptionParameters(index)
        columbic_force.addExclusion(iatom,jatom )
    
    system.addForce(columbic_force)


# Finally add all the particles to each force 
        
    for index in range(system.getNumParticles()):
        na_na_interaction_force.addParticle()
        cl_cl_interaction_force.addParticle()
        na_cl_interaction_force.addParticle()
        o_o_interaction_force.addParticle()
        na_o_interaction_force.addParticle()
        cl_o_interaction_force.addParticle()
    return system`

To run the system , use the following inside the directory

`python *.py`

Simulation Details-
System - NaCl + Water
Openmm version- 7.1.1
Openmm tools version - 0.14.0
Python version- Python 3.5.4

I will be more than happy if anyone can help/guide me in my toil :)

**Comment by peastman** [comment]
I think you're making the calculation much more complicated than it needs to be.  You have a really complicated combining rule involving 12 per-particle parameters.  Of course that's going to be slower than the standard force.  But do you really need to?  For example, here's your combining rule for sigma:

```
sigma = (sigma_nacl1+sigma_nacl2)/2 + (sigma_nao1+sigma_nao2)/2 + (sigma_clo1+sigma_clo2)/2 -
(sigma_nana1+sigma_nana2)/2 - (sigma_clcl1+sigma_clcl2)/2 - (sigma_oo1+sigma_oo2)/2;
```

Since that's just linear in all the terms, why don't you sum them to get a single sigma for each particle?  Then your combining rule could simplify to

```
sigma = sigma1+sigma2
```

> I would need to include both steric and electrostatic forces in CustomNonbondedForce (when I am removing the forces['NonbondedForce'] from my system)

Alternatively, just set epsilon to 0 for every particle in the NonbondedForce.  On the OpenCL and CUDA platforms, it will realize you aren't using LJ and ifdef out the parts of the kernel related to it.  This also works to a limited extent on the CPU platform.  Setting all the epsilons to 0 will reduce the time it spends on LJ, but it won't completely eliminate it the way it does on OpenCL or CUDA.

Also note that on the CPU platform, CustomNonbondedForce is always a little slower than NonbondedForce.  The calculation isn't as highly vectorized, and it does more of it in double precision.  On OpenCL or CUDA there should be very little difference between them (assuming, of course, that they're actually doing exactly the same calculation).

In your implementation of Coulomb, I notice you multiply the interaction by `erfc(alpha_ewald*r)`.  That's only appropriate if you're doing PME.  It serves to remove the part of the interaction that was computed in reciprocal space.  If you're actually doing that, you need to also include a NonbondedForce and have it compute the reciprocal space interactions.  It's possible to do that with force groups and a CustomIntegrator, but it's complicated.  If you're doing PME, it's much simpler to just let NonbondedForce handle all of the Coulomb interactions.

**Comment by arjunwadhawan** [comment]
Thanks for the reply @peastman .  I agree with you on most of the points.
**1**- I actually implemented the combination rule wrongly. It has to be as following to do what I wanted it to do. It is a bit complicate because this way I could get per particle parameters from single equation based on the values given in the table 

An example : _(Just for reference )_
_σ, ε_ for Na - Na => sigma = 0.2217 ; epsilon = 1.47236
_σ, ε_ for Cl - Cl  =>  sigma =0.4849 ; epsilon = 0.076923
But 
_σ, ε_ for Na - Cl => sigma = 0.2905 ; epsilon = 1.43889
which is neither direct arithmetic [(0.2217+0.4849)/2 = 0.3533 ] not geometric [sqrt(0.2217*0.4849) =0.3278] combination. So I wrote it as a liner combination of other values. (The reason for this is charge based scaling of these values as explained in the paper) 

``` python
sigma = sqrt(sigma_nacl1*sigma_nacl2) + sqrt(sigma_nao1*sigma_nao2) + sqrt(sigma_clo1*sigma_clo2) - sqrt(sigma_nana1*sigma_nana2) - sqrt(sigma_clcl1*sigma_clcl2) - sqrt(sigma_oo1*sigma_oo2);"
```
**2**- CPU thing was right. I tested on GPU and there the speed difference is 20 %.  Thanks for that 

**3**- Coulombic interaction part is what I was not aware of. Point noted. 

I am going to close the issue .

**Comment by peastman** [comment]
An alternative approach you could use is a lookup table for the parameters.  Have a single per-particle parameter called `type`, which is an integer index of the atom type.  Then create a pair of Discrete2DFunctions containing the sigma and epsilon values for each pair of types.  In place of the combining rule you would write

```
sigma = sigmaTable(type1, type2)
```

In this case, that might end up being faster.  Since you only have three atom types, the tables will be tiny and will fit easily in cache.
